[UX] Hint user when changing `spot.controller.resources` in skypilot config

cblmemo opened the pull request:
<!-- Describe the changes in this PR -->

Current spot implementation doesn't support modifying spot controller resources once the controller is launched. This PR adds a user hint for such cases.

Previously:
```bash
Traceback (most recent call last):
  File "/home/memory/install/miniconda3/envs/sky-dev/bin/sky", line 8, in <module>
    sys.exit(cli())
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/home/memory/skypilot/sky/utils/common_utils.py", line 311, in _record
    return f(*args, **kwargs)
  File "/home/memory/skypilot/sky/cli.py", line 1162, in invoke
    return super().invoke(ctx)
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/memory/skypilot/sky/utils/common_utils.py", line 311, in _record
    return f(*args, **kwargs)
  File "/home/memory/skypilot/sky/cli.py", line 1162, in invoke
    return super().invoke(ctx)
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/home/memory/skypilot/sky/utils/common_utils.py", line 332, in _record
    return f(*args, **kwargs)
  File "/home/memory/skypilot/sky/utils/common_utils.py", line 332, in _record
    return f(*args, **kwargs)
  File "/home/memory/skypilot/sky/cli.py", line 3740, in spot_launch
    retry_until_up=retry_until_up)
  File "/home/memory/skypilot/sky/utils/common_utils.py", line 332, in _record
    return f(*args, **kwargs)
  File "/home/memory/skypilot/sky/execution.py", line 755, in spot_launch
    retry_until_up=True,
  File "/home/memory/skypilot/sky/execution.py", line 320, in _execute
    retry_until_up=retry_until_up)
  File "/home/memory/skypilot/sky/utils/common_utils.py", line 332, in _record
    return f(*args, **kwargs)
  File "/home/memory/skypilot/sky/utils/common_utils.py", line 311, in _record
    return f(*args, **kwargs)
  File "/home/memory/skypilot/sky/backends/backend.py", line 57, in provision
    cluster_name, retry_until_up)
  File "/home/memory/skypilot/sky/backends/cloud_vm_ray_backend.py", line 2804, in _provision
    task, to_provision, cluster_name, dryrun)
  File "/home/memory/skypilot/sky/utils/common_utils.py", line 332, in _record
    return f(*args, **kwargs)
  File "/home/memory/skypilot/sky/backends/cloud_vm_ray_backend.py", line 4323, in _check_existing_cluster
    self.check_resources_fit_cluster(handle, task)
  File "/home/memory/skypilot/sky/backends/cloud_vm_ray_backend.py", line 2757, in check_resources_fit_cluster
    'Requested resources do not match the existing cluster.\n'
sky.exceptions.ResourcesMismatchError: Requested resources do not match the existing cluster.
  Requested:    1x AWS(cpus=4, disk_size=50) 
  Existing:     1x GCP(n2-standard-4, disk_size=50)
To fix: specify a new cluster name, or down the existing cluster first: sky down sky-spot-controller-402b1bba
```
Now:
```bash
sky.exceptions.ResourcesMismatchError: Requested resources do not match the existing cluster.
  Requested:    1x AWS(cpus=4, disk_size=50) 
  Existing:     1x GCP(n2-standard-4, disk_size=50)
To fix: specify a new cluster name, or down the existing cluster first: sky down sky-spot-controller-402b1bba

The above exception was the direct cause of the following exception:

RuntimeError: Modifying spot controller resources is not supported. Please change the spot.controller.resources in ~/.sky/config.yaml file to the original values or remove it instead.
```

<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
     - launch a spot controller with `cloud: gcp`; specify `cloud: aws` and `sky spot launch` again
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

concretevitamin commented on sky/execution.py:
```diff
@@ -745,15 +745,23 @@ def spot_launch(
-            retry_until_up=True,
-        )
+        try:
+            _execute(
```
Code gardening: document Raises: in _execute()'s docstr? (or at least add ResourcesMismatchError there)

concretevitamin commented on sky/execution.py:
```diff
@@ -745,15 +745,23 @@ def spot_launch(
+        except exceptions.ResourcesMismatchError as e:
+            with ux_utils.print_exception_no_traceback():
+                raise RuntimeError(
+                    'Modifying spot controller resources is not supported. '
+                    'Please change the spot.controller.resources in '
+                    '~/.sky/config.yaml file to the original values or '
+                    'remove it instead.') from e
```
How about:
```
Found existing spot controller but ~/.sky/config.yaml specifies a different spot.controller.resources field.\n
To change the spot controller's resources, `sky down` it first and redo `sky spot launch` for the configured resources to take effect.
```

concretevitamin reviewed (commented):
Nice @cblmemo!

Nit: I tried to spot launch with 
```
spot:
  controller:
    resources:  # same specs as 'resources' in a task YAML
      cloud: aws
      # region: us-central1
      # cpus: 2+  # number of vCPUs, max concurrent spot jobs = 2 * cpus
      disk_size: 50
```
first. Then I changed `disk_size` to 150 and reran `spot launch`. It went through without raising this error. Is it expected?

cblmemo commented:
> Nice @cblmemo!
> 
> Nit: I tried to spot launch with
> 
> ```
> spot:
>   controller:
>     resources:  # same specs as 'resources' in a task YAML
>       cloud: aws
>       disk_size: 50
> ```
> 
> first. Then I changed `disk_size` to 150 and reran `spot launch`. It went through without raising this error. Is it expected?
> 
> Likewise I added `cpus: 2` and expected an error but didn't see one (existing spot controller is m6i.2xlarge with 8 vCPUs).

This is the default behavior if we execute on an existing cluster (i.e. ignore CPU, memory, and disk size requirements.). It is implemented in the `sky.Resources.less_demanding_than`. Do you think we need to do more checks?

concretevitamin commented:
We may need to do something different than that func, because several users have wanted to change disk size / cpus of the spot controller, so this silent behavior may be surprising.

cblmemo commented:
> We may need to do something different than that func, because several users have wanted to change disk size / cpus of the spot controller, so this silent behavior may be surprising.

Good point! I'm thinking in such cases, should we print a warning or throw an error if the requested resources does not exactly match the controller resources? e.g. previously launched with 4 CPUs and now request 2 CPUs

concretevitamin commented:
Actually, one proposal is to change `less_demanding_than` to check  CPU, memory, and disk size properly. This seems orthogonal to ignoring them as scheduling requirements. Wdyt? cc @Michaelvll too
